src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,358 @@
+        {
+            EnsureGlobalLockIsOwned();
+            return _assemblies.Remove(assemblyFileName, out value);
+        }
+
+        public int AssemblyHandlersCount
+        {
+            get
+            {
+                EnsureGlobalLockIsOwned();
+                return _assemblies.Count;
+            }
+        }
+
+        public ValueTask<RegisterExtensionResponse> LoadAssemblyAsync(string assemblyFileName)
+        {
+            var assemblyFilePath = Path.Combine(AssemblyFolderPath, assemblyFileName);
+
+            // _extensionMessageHandlerService.RegisterAssembly locks _lockObject.
+            // You can lock _lockObject when holding a lock on AssemblyLoadLockObject, not vice-versa
+            if (Monitor.IsEntered(_extensionMessageHandlerService._lockObject))
+            {
+                throw new InvalidOperationException("Global lock should not be owned");
+            }
+
+            // AssemblyLoadLockObject is only used to avoid multiple calls from the same extensions to load the same assembly concurrently
+            // resulting in the constructors of the same handlers being called more than once.
+            // All other concurrent operations, including modifying extension.Assemblies are protected by _lockObject.
+            lock (_assemblyLoadLockObject)
+            {
+                AssemblyHandlers? assemblyHandlers = null;
+
+                try
+                {
+                    var assembly = AnalyzerAssemblyLoader.LoadFromPath(assemblyFilePath);
+                    var messageWorkspaceHandlers = _extensionMessageHandlerService._customMessageHandlerFactory.CreateWorkspaceMessageHandlers(assembly)
+                        .ToImmutableDictionary(h => h.Name, h => h);
+                    var messageDocumentHandlers = _extensionMessageHandlerService._customMessageHandlerFactory.CreateDocumentMessageHandlers(assembly)
+                        .ToImmutableDictionary(h => h.Name, h => h);
+
+                    assemblyHandlers = new AssemblyHandlers()
+                    {
+                        WorkspaceMessageHandlers = messageWorkspaceHandlers,
+                        DocumentMessageHandlers = messageDocumentHandlers,
+                    };
+
+                    // We don't add assemblyHandlers to _assemblies here and instead let _extensionMessageHandlerService.RegisterAssembly do it
+                    // since RegisterAssembly can still fail if there are duplicated handler names.
+                }
+                catch (Exception e) when (FatalError.ReportAndPropagate(e, ErrorSeverity.General))
+                {
+                    // unreachable
+                }
+                finally
+                {
+                    // In case of exception, we cache null so that we don't try to load the same assembly again.
+                    _extensionMessageHandlerService.RegisterAssembly(this, assemblyFileName, assemblyHandlers);
```

Comment:
i don't really see how the code follows the the above.  in the case of exception it seems like we might still have a non-null assemblyHandlers instance.  or, it could be easy to get that if someone just adds a single line after line 311.

in this case, you can assign e to a local exception, and if that exists here, pass null through as assemblyHandlers.

Reply:
The idea is that `assemblyHandlers` is set as the very last statement of the `catch` block, so it can only be non-null if there wasn't any exception.
But I did what you suggested and created an exception variable to identify a failure.